File: scripts/build-games-json.py
#!/usr/bin/env python3
"""Build public/data/games.json from data/games/*.yaml for the frontend."""

from __future__ import annotations

import json
import sys
from datetime import date, datetime
from pathlib import Path

try:
    import yaml
except ImportError:
    print("请安装 PyYAML: pip install pyyaml", file=sys.stderr)
    sys.exit(1)

ROOT = Path(__file__).resolve().parent.parent
GAMES_DIR = ROOT / "data" / "games"
OUT_DIR = ROOT / "web" / "public" / "data"
OUT_FILE = OUT_DIR / "games.json"
SRC_OUT = ROOT / "web" / "src" / "data" / "games.json"


def normalize(obj):
    if isinstance(obj, (date, datetime)):
        return obj.isoformat()
    if isinstance(obj, dict):
        return {k: normalize(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [normalize(v) for v in obj]
    return obj


def load_games() -> list[dict]:
    games: list[dict] = []
    files = sorted(GAMES_DIR.glob("*.yaml")) + sorted(GAMES_DIR.glob("*.yml"))
    for path in files:
        try:
            data = yaml.safe_load(path.read_text(encoding="utf-8"))
        except Exception as e:
            print(f"跳过 {path.name}: {e}", file=sys.stderr)
            continue
        if not isinstance(data, dict) or not data.get("id"):
            print(f"跳过 {path.name}: 无效数据", file=sys.stderr)
            continue
        data.pop("notes", None)
        games.append(normalize(data))
    games.sort(key=lambda g: g.get("updated_at") or g.get("release_date") or "", reverse=True)
    return games


def main() -> None:
    if not GAMES_DIR.exists():
        print(f"目录不存在: {GAMES_DIR}", file=sys.stderr)
        sys.exit(1)

    games = load_games()
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    SRC_OUT.parent.mkdir(parents=True, exist_ok=True)

    payload = json.dumps(games, ensure_ascii=False, indent=2)
    OUT_FILE.write_text(payload + "\n", encoding="utf-8")
    SRC_OUT.write_text(payload + "\n", encoding="utf-8")
    print(f"已写入 {len(games)} 条游戏 → {OUT_FILE.relative_to(ROOT)}")
    print(f"已写入 {len(games)} 条游戏 → {SRC_OUT.relative_to(ROOT)}")


if __name__ == "__main__":
    main()

File: scripts/build-games-json.mjs
#!/usr/bin/env node
/**
 * Build web/src/data/games.json + web/public/data/games.json from data/games/*.yaml
 * Pure Node — no Python/PyYAML required on Cloudflare Pages.
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import yaml from "js-yaml";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, "..");
const GAMES_DIR = path.join(ROOT, "data", "games");
const OUT_PUBLIC = path.join(ROOT, "web", "public", "data", "games.json");
const OUT_SRC = path.join(ROOT, "web", "src", "data", "games.json");

function loadGames() {
  if (!fs.existsSync(GAMES_DIR)) {
    console.error("missing", GAMES_DIR);
    process.exit(1);
  }
  const files = fs
    .readdirSync(GAMES_DIR)
    .filter((f) => /\.ya?ml$/i.test(f))
    .sort();

  const games = [];
  for (const file of files) {
    const full = path.join(GAMES_DIR, file);
    try {
      const raw = fs.readFileSync(full, "utf8");
      const data = yaml.load(raw);
      if (!data || typeof data !== "object" || !data.id) {
        console.warn("skip invalid", file);
        continue;
      }
      delete data.notes;
      games.push(data);
    } catch (e) {
      console.warn("skip", file, e.message);
    }
  }

  games.sort((a, b) => {
    const da = a.updated_at || a.release_date || "";
    const db = b.updated_at || b.release_date || "";
    return String(db).localeCompare(String(da));
  });
  return games;
}

const games = loadGames();
const payload = JSON.stringify(games, null, 2) + "\n";

fs.mkdirSync(path.dirname(OUT_PUBLIC), { recursive: true });
fs.mkdirSync(path.dirname(OUT_SRC), { recursive: true });

// Windows 下已存在的产物文件可能被文件监视器/杀软占用写锁，
// 先 unlink 再写新文件可稳定绕过 EPERM。
function safeWrite(target, content) {
  try {
    fs.unlinkSync(target);
  } catch {
    /* 文件不存在或被占用时忽略，下面 writeFileSync 会报错 */
  }
  fs.writeFileSync(target, content, "utf8");
}

safeWrite(OUT_PUBLIC, payload);
safeWrite(OUT_SRC, payload);
console.log(`synced ${games.length} games → web/src/data/games.json + public`);
